pkg/plugin/aggregation/run_test.go

```diff
@@ -0,0 +1,198 @@
+		t.Fatalf("Could not generate test cert: %v", err)
+	}
+	after = func(time.Duration) <-chan time.Time { return time.After(250 * time.Millisecond) }
+
+	testCases := []struct {
+		desc string
+
+		expectNumResults   int
+		expectStillRunning bool
+		forceResults       bool
+		cancelContext      bool
+
+		plugin           plugin.Interface
+		expectedResults  []plugin.ExpectedResult
+		podList          *corev1.PodList
+		podCreationError error
+	}{
+		{
+			desc:            "Continue monitoring if no results/errors",
+			plugin:          testPlugin,
+			expectedResults: testPluginExpectedResults,
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{healthyPod},
+			},
+			expectStillRunning: true,
+		}, {
+			desc:            "Error launching plugin causes exit and plugin result",
+			plugin:          testPlugin,
+			expectedResults: testPluginExpectedResults,
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{healthyPod},
+			},
+			podCreationError: errors.New("createPod error"),
+			expectNumResults: 1,
+		}, {
+			desc:            "Failing plugin causes exit and plugin result",
+			plugin:          testPlugin,
+			expectedResults: testPluginExpectedResults,
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{failingPod},
+			},
+			expectNumResults: 1,
+		}, {
+			desc:            "Plugin obtaining results in exits",
+			plugin:          testPlugin,
+			expectedResults: testPluginExpectedResults,
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{healthyPod},
+			},
+			forceResults:     true,
+			expectNumResults: 1,
+		}, {
+			desc:            "Context cancellation results in exit",
+			plugin:          testPlugin,
+			expectedResults: testPluginExpectedResults,
+			podList: &corev1.PodList{
+				Items: []corev1.Pod{healthyPod},
+			},
+			cancelContext:    true,
+			expectNumResults: 0,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			a := NewAggregator(".", tc.expectedResults)
+			ctx, cancel := context.WithCancel(context.Background())
+
+			fclient := fake.NewSimpleClientset()
+			fclient.PrependReactor("list", "pods", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+				return true, tc.podList, nil
+			})
+			fclient.PrependReactor("create", "pods", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+				return true, nil, tc.podCreationError
+			})
+
+			doneCh, timeoutCh := make(chan (struct{}), 1), make(chan (struct{}), 1)
+			if tc.cancelContext {
+				cancel()
+			} else {
+				// Max timeout for test to unblock.
+				go func() {
+					time.Sleep(2 * time.Second)
+					timeoutCh <- struct{}{}
+					cancel()
+				}()
+			}
+
+			go func() {
+				a.RunAndMonitorPlugin(ctx, tc.plugin, fclient, nil, "testname", testCert)
+				doneCh <- struct{}{}
+			}()
+
+			if tc.forceResults {
+				a.resultsMutex.Lock()
+				a.Results["myPlugin"] = &plugin.Result{}
+				a.resultsMutex.Unlock()
+			}
+
+			if tc.cancelContext {
```

Comment:
Is it necessary to cancel this context again?

Reply:
Thanks for seeing that I had done it twice.